packages/app-utils/src/components/ConnectionBootstrap.tsx

```diff
@@ -56,47 +59,63 @@ export function ConnectionBootstrap({
     [api, client]
   );
 
-  useEffect(
-    function listenForShutdown() {
-      if (connection == null) return;
+  useEffect(() => {
+    if (connection == null) return;
 
-      function handleShutdown(event: CustomEvent): void {
-        const { detail } = event;
-        log.info('Shutdown', `${JSON.stringify(detail)}`);
-        setError(`Server shutdown: ${detail ?? 'Unknown reason'}`);
-      }
+    // handles the shutdown event
+    function handleShutdown(event: CustomEvent): void {
+      const { detail } = event;
+      log.info('Shutdown', `${JSON.stringify(detail)}`);
+      setError(`Server shutdown: ${detail ?? 'Unknown reason'}`);
+    }
+    const removerFn = connection.addEventListener(
+      api.IdeConnection.EVENT_SHUTDOWN,
+      handleShutdown
+    );
 
-      const removerFn = connection.addEventListener(
-        api.IdeConnection.EVENT_SHUTDOWN,
-        handleShutdown
+    // handles the auth failed event
+    function handleAuthFailed(event: CustomEvent): void {
+      const { detail } = event;
+      log.warn('Reconnect authentication failed', `${JSON.stringify(detail)}`);
+      setError(
+        `Reconnect authentication failed: ${detail ?? 'Unknown reason'}`
       );
-      return removerFn;
-    },
-    [api, connection]
-  );
+      setIsAuthFailedState(true);
+    }
+    const authFailed = connection.addEventListener(
+      dh.CoreClient.EVENT_RECONNECT_AUTH_FAILED,
+      handleAuthFailed
+    );
+
+    return authFailedState ? authFailed : removerFn;
```

Comment:
These should be separate effects. The return here is not correct.

The return function of an effect is run when the effect is changed before it runs the next version of the effect. So in this case if the connection changes we are removing the old listener by returning `removerFn` and then attaching a new listener. This implementation would only remove 1 of the event listeners and leave the other attached.

Also, effects should generally only cover 1 thing/lifecycle so they are more clear.